components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodClientProjectSessionTracker.kt

```diff
@@ -35,6 +58,140 @@ class GitpodClientProjectSessionTracker(
         }
     }
 
+    private fun isExposedServedPort(port: Status.PortsStatus?) : Boolean {
+        if (port === null) {
+            return false
+        }
+        return port.served && port.hasExposed()
+    }
+
+    private fun showOpenServiceNotification(port: PortsStatus, offerMakePublic: Boolean = false) {
+        val message = "A service is available on port ${port.localPort}"
+        val notification = manager.notificationGroup.createNotification(message, NotificationType.INFORMATION)
+
+        val openBrowserAction = NotificationAction.createSimple("Open Browser") {
+            openBrowser(port.exposed.url)
+        }
+        notification.addAction(openBrowserAction)
+
+        if (offerMakePublic) {
+            val makePublicLambda = {
+                runBlocking {
+                    makePortPublic(info.workspaceId, port)
+                }
+            }
+            val makePublicAction = NotificationAction.createSimple("Make Public", makePublicLambda)
+            notification.addAction(makePublicAction)
+        }
+
+        ClientId.withClientId(session.clientId) {
+            val project = RestService.getLastFocusedOrOpenedProject()
+            notification.notify(project)
+        }
+    }
+
+    private suspend fun makePortPublic(workspaceId: String, port: PortsStatus) {
+        val p = WorkspaceInstancePort()
+        p.port = port.localPort
+        p.visibility = io.gitpod.gitpodprotocol.api.entities.PortVisibility.PUBLIC.toString()
+        p.url = port.exposed.url
+
+        try {
+            manager.client.server.openPort(workspaceId, p).await()
```

Comment:
at the moment, pressing the button doesn't update the UI (in VS Code in the ports view you can see the text "private" change into "public"), so it looks a bit unfinished. We could consider adding an extra notification like:

```kotlin
val message = "The port ${port.localPort} is now public"
val notification = manager.notificationGroup.createNotification(message, NotificationType.INFORMATION)
ClientId.withClientId(session.clientId) {
   val project = RestService.getLastFocusedOrOpenedProject()
   notification.notify(project)
}
```

Reply:
Too many notification is also not nice. Is it possible to update an action? Maybe just make it expirable?